packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -479,6 +478,103 @@ void Add(
   }
 }
 
+namespace ColMapFunctors
+{
+  template<typename ByteView, typename GView>
+  struct UnionEntries
+  {
+    UnionEntries(ByteView& entryUnion_, const GView& gids_) : entryUnion(entryUnion_), gids(gids_) {}
+    KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const
+    {
+      entryUnion(gids(i)) = 1;
+    }
+    ByteView& entryUnion;
+    const GView& gids;
+  };
+
+  template<typename ByteView, typename LView>
+  struct EntriesPrefixSum
+  {
+    EntriesPrefixSum(const ByteView& entryUnion_, LView& gtol_) : entryUnion(entryUnion_), gtol(gtol_) {}
+    KOKKOS_INLINE_FUNCTION void operator()(const size_t i, size_t& update, const bool finalPass) const
+    {
+      // exclusive scan
+      if(finalPass)
+      {
+        gtol(i) = update;
+      }
+      if(entryUnion(i))
+      {
+        update++;
+      }
+      if(finalPass)
+      {
+        if(i == entryUnion.dimension_0() - 1)
+        {
+          gtol(i + 1) = update;
+        }
+      }
+    }
+    const ByteView& entryUnion;
+    LView& gtol;
+  };
+
+  template<typename LView, typename GView>
+  struct ConvertGlobalToLocal
+  {
+    ConvertGlobalToLocal(const LView& gtol_, const GView& gids_, LView& lids_) : gtol(gtol_), gids(gids_), lids(lids_) {}
+    KOKKOS_INLINE_FUNCTION void operator()(const size_t i) const
+    {
+      lids(i) = gtol(gids(i));
+    }
+    const LView& gtol;
```

Comment:
Kokkos::View has shallow copy semantics.  If your functor holds onto Kokkos::View instances by nonconst reference, this functor won't work on GPUs.  A reference is a pointer, and it points to host memory, where the class instance lives.  The class instance _contains_ a GPU device allocation.  Just store the Kokkos::View by value in the functor.  Also, "const View<T>" doesn't mean the same thing as "View<const T>".  You'll have to store these as "View" rather than "const View" anyway.